tools/ExtensionsMetadataGenerator/src/ExtensionsMetadataGenerator.Console/ExtensionsMetadataGenerator.cs

```diff
@@ -81,44 +74,97 @@ public static string GenerateExtensionsJson(IEnumerable<ExtensionReference> exte
+            BaseAssemblyResolver resolver = new DefaultAssemblyResolver();
+            resolver.AddSearchDirectory(Path.GetDirectoryName(fileName));
+
+            ReaderParameters readerParams = new ReaderParameters { AssemblyResolver = resolver };
+
+            AssemblyDefinition assembly = AssemblyDefinition.ReadAssembly(fileName, readerParams);
+
+            var startupAttributes = assembly.Modules.SelectMany(p => p.GetCustomAttributes())
+                .Where(a => IsWebJobsStartupAttributeType(a.AttributeType, logger));
 
             List<ExtensionReference> extensionReferences = new List<ExtensionReference>();
             foreach (var attribute in startupAttributes)
             {
-                var nameProperty = attribute.GetType().GetProperty("Name");
-                var typeProperty = attribute.GetType().GetProperty("WebJobsStartupType");
+                var typeProperty = attribute.ConstructorArguments.ElementAtOrDefault(0);
+                var nameProperty = attribute.ConstructorArguments.ElementAtOrDefault(1);
+
+                TypeDefinition typeDef = (TypeDefinition)typeProperty.Value;
+                string assemblyQualifiedName = Assembly.CreateQualifiedName(typeDef.Module.Assembly.FullName, GetReflectionFullName(typeDef));
+
+                string name = GetName((string)nameProperty.Value, typeDef);
 
                 var extensionReference = new ExtensionReference
                 {
-                    Name = (string)nameProperty.GetValue(attribute),
-                    TypeName = ((Type)typeProperty.GetValue(attribute)).AssemblyQualifiedName
+                    Name = name,
+                    TypeName = assemblyQualifiedName
                 };
 
                 extensionReferences.Add(extensionReference);
             }
 
             return extensionReferences;
         }
+
+        // Because we're now using static analysis we can't rely on the constructor running. Copying the constructor logic from:
+        // https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsStartupAttribute.cs#L33-L47.
+        private static string GetName(string name, TypeDefinition startupTypeDef)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                // for a startup class named 'CustomConfigWebJobsStartup' or 'CustomConfigStartup',
+                // default to a name 'CustomConfig'
+                name = startupTypeDef.Name;
+                int idx = name.IndexOf("WebJobsStartup");
+                if (idx < 0)
+                {
+                    idx = name.IndexOf("Startup");
+                }
+                if (idx > 0)
+                {
+                    name = name.Substring(0, idx);
+                }
+            }
+
+            return name;
+        }
+
+        public static string GetReflectionFullName(TypeReference typeRef)
+        {
+            return typeRef.FullName.Replace("/", "+");
+        }
```

Comment:
This is the trickiest part. Cecil uses the ECMA standard for type names, which is different than what Reflection uses. FullName of nested types use "/" rather than "+".